File: Source/AlphaTab/Rendering/Glyphs/TrillGlyph.cs
using System;
using AlphaTab.Platform;

namespace AlphaTab.Rendering.Glyphs
{
    internal class TrillGlyph : EffectGlyph
    {
        public TrillGlyph(float x, float y)
            : base(x, y)
        {
        }

        public override void DoLayout()
        {
            base.DoLayout();
            Height = 20 * Scale;
        }

        public override void Paint(float cx, float cy, ICanvas canvas)
        {
            var res = Renderer.Resources;

            canvas.Font = res.MarkerFont;

            var textw = canvas.MeasureText("tr");
            canvas.FillText("tr", cx + X, cy + Y + canvas.Font.Size / 2);

            var startX = textw + 3 * Scale;
            var endX = Width - startX;
            var waveScale = 1.2f;
            var step = 11 * Scale * waveScale;
            var loops = Math.Max(1, (endX - startX) / step);


            var loopX = startX;
            var loopY = cy + Y + Height;
            for (var i = 0; i < loops; i++)
            {
                canvas.FillMusicFontSymbol(cx + X + loopX, loopY, waveScale, MusicFontSymbol.WaveHorizontalSlight);
                loopX += step;
            }
        }
    }
}

File: src/rendering/glyphs/TrillGlyph.ts
import type { ICanvas } from '@src/platform/ICanvas';
import { MusicFontSymbol } from '@src/model/MusicFontSymbol';
import { BeatXPosition } from '@src/rendering/BeatXPosition';
import { GroupedEffectGlyph } from '@src/rendering/glyphs/GroupedEffectGlyph';

export class TrillGlyph extends GroupedEffectGlyph {
    public constructor(x: number, y: number) {
        super(BeatXPosition.EndBeat);
        this.x = x;
        this.y = y;
    }

    public override doLayout(): void {
        super.doLayout();
        this.height = this.renderer.smuflMetrics.GlyphHeights.get(MusicFontSymbol.OrnamentTrill)! / 2;
    }

    protected override paintGrouped(cx: number, cy: number, endX: number, canvas: ICanvas): void {
        let startX: number = cx + this.x;

        canvas.fillMusicFontSymbol(startX, cy + this.y + this.height, 1, MusicFontSymbol.OrnamentTrill, true);

        startX += this.renderer.smuflMetrics.GlyphWidths.get(MusicFontSymbol.OrnamentTrill)! / 2;

        const waveScale: number = this.renderer.smuflMetrics.trillWaveScale;
        const step: number = this.renderer.smuflMetrics.GlyphWidths.get(MusicFontSymbol.WiggleTrill)! * waveScale;
        const loops: number = Math.floor((endX - startX) / step);
        const loopY: number = cy + this.y + this.height * this.renderer.smuflMetrics.trillLoopHeightToY;
        let loopX: number = startX;
        for (let i: number = 0; i < loops; i++) {
            canvas.fillMusicFontSymbol(loopX, loopY, waveScale, MusicFontSymbol.WiggleTrill, false);
            loopX += step;
        }
    }
}
